libp2p/autotls/acme/api.nim

```diff
@@ -0,0 +1,342 @@
+    chalURL: string,
+    key: KeyPair,
+    kid: Kid,
+    retries: int = DefaultChalCompletedRetries,
+): Future[void] {.async: (raises: [ACMEError, CancelledError]).} =
+  let emptyPayload = newJObject()
+
+  let completedResponse = wrapSerialization:
+    (await self.signedAcmeRequest(chalURL, emptyPayload, key, kid = Opt.some(kid))).body.to(
+      ACMECompletedResponse
+    )
+
+  # check until acme server is done (poll validation)
+  for _ in 0 .. retries:
+    let rawResponse =
+      try:
+        await HttpClientRequestRef
+        .get(self.session, completedResponse.checkURL)
+        .get()
+        .send()
+      except HttpError as exc:
+        raise newException(ACMEError, "Failed to connect to ACME server", exc)
+      except ValueError as exc:
+        raise newException(ACMEError, "Failed to parse JSON", exc)
+    let checkResponse = wrapSerialization:
+      (await rawResponse.getResponseBody()).to(ACMECheckResponse)
+
+    case checkResponse.status
+    of "pending":
+      var retryAfter: Duration
+      try:
+        retryAfter = parseInt(rawResponse.headers.getString("Retry-After")).seconds
+      except ValueError:
+        retryAfter = DefaultChalCompletedRetryTime
+      await sleepAsync(retryAfter) # try again after some delay
+    of "valid":
+      return
+    else:
+      raise newException(
+        ACMEError,
+        "Failed challenge completion: expected 'valid', got '" & checkResponse.status &
+          "'",
+      )
+
+proc finalizeCertificate*(
+    self: ACMEApi,
+    domain: string,
+    finalizeURL: string,
+    orderURL: string,
+    key: KeyPair,
+    kid: Kid,
+    retries: int = DefaultFinalizeRetries,
+): Future[bool] {.async: (raises: [ACMEError, CancelledError]).} =
+  var certKey: cert_key_t
+  var certCtx: cert_context_t
+  var derCSR: ptr cert_buffer = nil
+
+  let personalizationStr = "libp2p_autotls"
+  if cert_init_drbg(
+    personalizationStr.cstring, personalizationStr.len.csize_t, certCtx.addr
+  ) != CERT_SUCCESS:
+    raise newException(ACMEError, "Failed to initialize certCtx")
+  if cert_generate_key(certCtx, certKey.addr) != CERT_SUCCESS:
+    raise newException(ACMEError, "Failed to generate cert key")
+
+  if cert_signing_req(domain.cstring, certKey, derCSR.addr) != CERT_SUCCESS:
+    raise newException(ACMEError, "Failed to create CSR")
+
+  let b64CSR = base64.encode(derCSR.toSeq, safe = true)
+  let payload = %*{"csr": b64CSR}
+
+  # keep checking order until it's finalized
+  var checkResponse: ACMECheckResponse
+  for _ in 0 .. retries:
+    let finalizedResponse = wrapSerialization:
+      (await self.signedAcmeRequest(finalizeURL, payload, key, kid = Opt.some(kid))).body.to(
+        ACMEFinalizedResponse
+      )
+    let rawResponse =
+      try:
+        await HttpClientRequestRef.get(self.session, orderURL).get().send()
+      except CatchableError as exc:
+        raise
+          newException(ACMEError, "Unexpected error while finalizing certificate", exc)
```

Comment:
error here  is probably not correctly handled -> if `CancelledError` happens here it will be convered to `ACMEError`, which is not what we really want.
whenever there is `await` within `try` you need to handle `CancelledError` separately and re-raise this exception without converting it. additional you may need to preform cleanups related current method call (if any).